packages/repository/src/define-repository-class.ts

```diff
@@ -0,0 +1,156 @@
+// Copyright IBM Corp. 2020. All Rights Reserved.
+// Node module: @loopback/repository
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Constructor} from '@loopback/context';
+import assert from 'assert';
+import {Entity, Model} from './model';
+import {
+  CrudRepository,
+  DefaultCrudRepository,
+  DefaultKeyValueRepository,
+  EntityCrudRepository,
+  juggler,
+  KeyValueRepository,
+} from './repositories';
+
+/**
+ * Signature for CrudRepository classes
+ *
+ * @typeParam M - Model class
+ * @typeParam R - Repository class/interface
+ */
+export interface CrudRepositoryClass<
+  M extends Model,
+  R extends CrudRepository<M>
+> {
+  new (ds: juggler.DataSource): R;
```

Comment:
A typical "base" repository class requires two arguments:
- The datasource to use
- The model class to persist

Custom base repository implementations may require more arguments, e.g. a third `options` argument.

Reply:
OK, now I see that you moved this interface from my original implementation.

It's important to keep in mind that in the original version, this interface was used to describe the type returned by the factory function, it's a repository bound to a given model.

When reviewing the code below, I got confused because this distinction was not clear. In other places, we use `FooClass` as an alias for `typeof Foo`, but here `CrudRepositoryClass` is not necessarily the same thing as `typeof CrudRepository`.

Can we please find a name for this interface which will better communicate the fact that it's describing a class returned by one of the factory functions below?